crates/swc_ecma_transforms_base/src/dropped_comments_preserver.rs

```diff
@@ -0,0 +1,55 @@
+use swc_common::{comments::Comments, BytePos, Span};
+use swc_ecma_visit::{as_folder, noop_visit_mut_type, Fold, VisitMut, VisitMutWith};
+/// Preserves comments that would otherwise be dropped.
+///
+/// If during compilation an ast node associated with
+/// a comment is dropped, the comment will not appear in the final emitted
+/// output. This can create problems in the JavaScript ecosystem, particularly
+/// around instanbul coverage and other tooling that relies on comment
+/// directives.
+///
+/// This transformer shifts orphaned comments to the next closest ast-node.
+
+pub fn dropped_comments_preserver(comments: Option<&dyn Comments>) -> impl '_ + Fold + VisitMut {
+    as_folder(DroppedCommentsPreserver {
+        comments,
+        comment_position: 0,
+        is_first_span: true,
+    })
+}
+
+struct DroppedCommentsPreserver<'a> {
+    comments: Option<&'a dyn Comments>,
+    comment_position: u32,
+    is_first_span: bool,
+}
+
+impl VisitMut for DroppedCommentsPreserver<'_> {
+    noop_visit_mut_type!();
+
+    fn visit_mut_span(&mut self, span: &mut Span) {
+        if span.is_dummy() || self.is_first_span {
```

Comment:
When rearranging comments, we have to stop and think what we should do with the module / script span. A comment at the top could be dropped, could be preserved in place, or could be transported to the next available span. Example of this ambiguity:

On the one hand, I could have something like this:

```typescript
/* Some doc block, this very clearly should remain at the top of the file */
type X = number;
...
```

Or I could have something like this:

```
/* istanbul ignore next-line */
type X = number;
const somethingFancy = (...x)  => x;
```

Because during transpilation code such as the second example will be shifted down to support utility functions, I opted to solve this ambiguity by assuming that the comment _must_ be moved to the next nearest existing span. This way at the very minimum tooling is supported.